### DISCUSSION TITLE
How would you go about rewrites in middleware with retries?

### DISCUSSION THREAD

**Discussion by neoromantic** [vercel/next.js#71487] [Help]
### Summary

I use middleware to proxy posthog ingest entrypoint as they state in their documentation: https://posthog.com/docs/advanced/proxy/nextjs

I've noticed that sometimes (quite often!) Posthog responds with server error (503), so I'd like to be able to retry such requests. 

Currently, code is simple:

```
    const url = request.nextUrl.clone()
    const hostname = url.pathname.startsWith("/ingest/static/")
      ? "us-assets.i.posthog.com"
      : process.env.NEXT_PUBLIC_POSTHOG_HOST!.replace("https://", "")

    const requestHeaders = new Headers(request.headers)
    requestHeaders.set("host", hostname)

    url.protocol = "https"
    url.hostname = hostname
    url.port = "443"
    url.pathname = url.pathname.replace(/^\/ingest/, "")

    return NextResponse.rewrite(url, {
      headers: requestHeaders,
    })
```

How would you go about making retries here, maybe with fetch-retry or whatever. 

Is there any reasonable solution to reuse NextResponse.rewrite, but taking care of occasional remote server errors?

**Comment by victorcesae** [comment]
I suggest handling the retry on the client side while the middleware only handle the proxy request.

**Reply by neoromantic** [reply]
Yeah, but on client side it's 3rd party code (posthog's), I can't change it.

**Comment by icyJoseph** [comment]
I think the problem here is that, the browser receives this redirect, follows it, and gets the 503. At this point the middleware has already stopped, so yeah, you'd need to know on the client side, that you did get a 503, and retry. 

Is there any way for you to listen to errors on the posthog client, and possibly, check what type of error, and retry? I googled around and apparently their JS SDK does retry though.

**Reply by icyJoseph** [reply]
Also, I'd look more into why it 503's so often. I found some posthog docs commenting that, hosting in Heroku and next.config.js rewrites could cause that, but middleware was presented as a solution! 

- https://posthog.com/docs/advanced/proxy/nextjs

<img width="1275" alt="Screenshot 2024-10-21 at 20 50 23" src="https://github.com/user-attachments/assets/de67630c-fc50-47f7-bad6-eefbba74f5e4">
